sweep/sweeper.go

```diff
@@ -859,22 +687,62 @@ func (s *UtxoSweeper) removeExclusiveGroup(group uint64) {
 func (s *UtxoSweeper) sweepCluster(cluster inputCluster,
 	currentHeight int32) error {
 
-	// Execute the sweep within a coin select lock. Otherwise the coins that
-	// we are going to spend may be selected for other transactions like
-	// funding of a channel.
+	// Execute the sweep within a coin select lock. Otherwise the coins
+	// that we are going to spend may be selected for other transactions
+	// like funding of a channel.
 	return s.cfg.Wallet.WithCoinSelectLock(func() error {
-		// Examine pending inputs and try to construct
-		// lists of inputs.
-		inputLists, err := s.getInputLists(cluster, currentHeight)
+		// Examine pending inputs and try to construct lists of inputs.
+		allSets, newSets, err := s.getInputLists(cluster, currentHeight)
 		if err != nil {
-			return fmt.Errorf("unable to examine pending inputs: %v", err)
+			return fmt.Errorf("examine pending inputs: %w", err)
 		}
 
-		// Sweep selected inputs.
-		for _, inputs := range inputLists {
-			err := s.sweep(inputs, cluster.sweepFeeRate, currentHeight)
+		// errAllSets records the error from broadcasting the sweeping
+		// transactions for all input sets.
+		var errAllSets error
+
+		// allSets contains retried inputs and new inputs. To avoid
+		// creating an RBF for the new inputs, we'd sweep this set
+		// first.
+		for _, inputs := range allSets {
+			errAllSets = s.sweep(
```

Comment:
We only return an error when the publishing of the transaction failed if its neither a RBF-Failure nor a DoubleSpend. So I think this will not allow us to broadcast the Newset when RBF conditions are not met because we map a Replacement Err to a [doubleSpend](https://github.com/lightningnetwork/lnd/blob/master/lnwallet/btcwallet/btcwallet.go#L1211-L1216) hence [circumventing](https://github.com/lightningnetwork/lnd/blob/master/sweep/sweeper.go#L1359-L1361) the error.

Maybe as a short term fix before a more sophisticated sweeper, we always return an error when the transaction fails due to publishing and also increase the reattempt times for every input. I don't see any benefit treating unexpected errors of the publisher differently tbh.

Reply:
yeah right, forgot about that part! Maybe I should add an itest to validate the changes, and yes, think we should just return the error here.